File: pulsar-common/src/main/java/ai/platon/pulsar/common/ConcurrentLRUCache.java
package ai.platon.pulsar.common;

import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Function;

/**
 * A very simple yet fast LRU cache with TTL support
 */
public class ConcurrentLRUCache<K, V> {
    /**
     * A fast yet short life least recently used cache
     */
    private final LinkedHashMap<String, V> cache;
    /**
     * Expires in second
     */
    private long ttl;

    public ConcurrentLRUCache(int capacity) {
        this(0, capacity);
    }

    /**
     * Construct a least recently used cache
     *
     * @param ttl      Time to live for items
     * @param capacity The max size of the cache
     */
    public ConcurrentLRUCache(Duration ttl, int capacity) {
        this.ttl = ttl.getSeconds();
        this.cache = new LinkedHashMap<String, V>(capacity, 0.75F, true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry<String, V> eldest) {
                return size() > capacity;
            }
        };
    }

    /**
     * Construct a least recently used cache
     *
     * @param ttl      Time to live for items, in seconds
     * @param capacity The max size of the cache
     */
    public ConcurrentLRUCache(long ttl, int capacity) {
        this.ttl = ttl;
        this.cache = new LinkedHashMap<>(capacity, 0.75F, true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry<String, V> eldest) {
                return size() > capacity;
            }
        };
    }

    public long getTtl() {
        return ttl;
    }

    public V get(K key) {
        String ttlKey = getTTLKey(key);
        V v;
        synchronized (cache) {
            v = cache.get(ttlKey);
        }

        return v;
    }

    public void put(K key, V v) {
        String ttlKey = getTTLKey(key);
        synchronized (cache) {
            cache.put(ttlKey, v);
        }
    }

    public V remove(K key) {
        String ttlKey = getTTLKey(key);
        V v;
        synchronized (cache) {
            v = cache.remove(ttlKey);
        }
        return v;
    }

    public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
        String ttlKey = getTTLKey(key);
        V v;
        synchronized (cache) {
            v = cache.get(ttlKey);
            if (v == null) {
                v = mappingFunction.apply(key);
                cache.put(ttlKey, v);
            }
        }

        return v;
    }

    public void clear() {
        synchronized (cache) {
            cache.clear();
        }
    }

    private String getTTLKey(K key) {
        if (ttl <= 0) {
            return key.toString();
        }

        long secondsDivTTL = System.currentTimeMillis() / 1000 / ttl;
        return secondsDivTTL + "\t" + key.toString();
    }
}

File: pulsar-common/src/main/kotlin/ai/platon/pulsar/common/concurrent/ConcurrentLRUCache.kt
package ai.platon.pulsar.common.concurrent

import java.time.Duration
import java.util.*

/**
 * A very simple yet fast LRU cache with TTL support
 */
class ConcurrentLRUCache<K, V> {
    /**
     * A fast yet short life least recently used cache
     */
    private val cache: LinkedHashMap<String, V>
    /**
     * Expires in seconds
     */
    var ttl: Long
        private set

    constructor(capacity: Int) : this(0, capacity) {}
    /**
     * Construct a least recently used cache
     *
     * @param ttl      Time to live for items
     * @param capacity The max size of the cache
     */
    constructor(ttl: Duration, capacity: Int) {
        this.ttl = ttl.seconds
        cache = object : LinkedHashMap<String, V>(capacity, 0.75f, true) {
            override fun removeEldestEntry(eldest: Map.Entry<String, V>): Boolean {
                return size > capacity
            }
        }
    }

    /**
     * Construct a least recently used cache
     *
     * @param ttl      Time to live for items, in seconds
     * @param capacity The max size of the cache
     */
    constructor(ttl: Long, capacity: Int) {
        this.ttl = ttl
        cache = object : LinkedHashMap<String, V>(capacity, 0.75f, true) {
            override fun removeEldestEntry(eldest: Map.Entry<String, V>): Boolean {
                return size > capacity
            }
        }
    }

    operator fun get(key: K): V? {
        val ttlKey = getTTLKey(key)
        synchronized(cache) { return cache[ttlKey] }
    }

    fun put(key: K, v: V) {
        val ttlKey = getTTLKey(key)
        synchronized(cache) { cache.put(ttlKey, v) }
    }

    fun remove(key: K): V? {
        val ttlKey = getTTLKey(key)
        synchronized(cache) { return cache.remove(ttlKey) }
    }

    fun computeIfAbsent(key: K, mappingFunction: (K) -> V): V {
        val ttlKey = getTTLKey(key)
        synchronized(cache) {
            return cache[ttlKey] ?: mappingFunction(key).also { cache[ttlKey] = it }
        }
    }

    fun clear() = synchronized(cache) { cache.clear() }

    private fun getTTLKey(key: K): String {
        if (ttl <= 0) {
            return key.toString()
        }
        val secondsDivTTL = System.currentTimeMillis() / 1000 / ttl
        return secondsDivTTL.toString() + "\t" + key.toString()
    }
}
